modules/lib/hyprlang.nix

```diff
@@ -0,0 +1,44 @@
+{ lib }: rec {
+  toHyprconf = with lib;
+    attrs: indentLevel: sourceFirst:
+    let
+      indent = concatStrings (replicate indentLevel "  ");
+
+      sections = filterAttrs (n: v: isAttrs v && n != "device") attrs;
+
+      mkSection = n: attrs: ''
+        ${indent}${n} {
+        ${toHyprconf attrs (indentLevel + 1) sourceFirst}${indent}}
+      '';
+
+      mkDeviceCategory = device: ''
+        ${indent}device {
+          name=${device.name}
+        ${
+          toHyprconf (filterAttrs (n: _: "name" != n) device) (indentLevel + 1)
+          sourceFirst
+        }${indent}}
+      '';
+
+      deviceCategory = lib.optionalString (hasAttr "device" attrs)
+        (if isList attrs.device then
+          (concatMapStringsSep "\n" (d: mkDeviceCategory d) attrs.device)
+        else
+          mkDeviceCategory attrs.device);
+
+      mkFields = generators.toKeyValue {
+        listsAsDuplicateKeys = true;
+        inherit indent;
+      };
+      allFields = filterAttrs (n: v: !(isAttrs v) && n != "device") attrs;
+
+      importantFields = filterAttrs (n: _:
+        (hasPrefix "$" n) || (hasPrefix "bezier" n)
+        || (sourceFirst && (hasPrefix "source" n))) allFields;
```

Comment:
Maybe `importantFields` should be extracted into a function argument?

Reply:
@fufexan  should these be hardcoded in each module as an argument or would it be another config option from the module to pass to the library?

Reply:
Decided to make it an option... I could see a benefit to a user being able to organize their config themself through it.